mill-universal-packager/src/io/github/hoangmaihuy/mill/packager/universal/UniversalPackagerModule.scala

```diff
@@ -58,4 +88,117 @@ trait UniversalPackagerModule extends PackagerModule {
+    Archiver().mkTarball(staged.path, out, universalTopLevelPath())(new GzipCompressorOutputStream(_, parameter))
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
+
+  /** Create a .tar.bz2 package file. The task will run universalStage() first, and then tarball and compress the stage
+    * output directory as the result.
+    */
+  def universalStagePackageTarBzip2: T[PathRef] = T {
+    val staged: PathRef = universalStage()
+    val out: Path = T.dest / (packageName() + universalTarBzip2Ext())
+    Archiver().mkTarball(staged.path, out, universalTopLevelPath())(new BZip2CompressorOutputStream(_))
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
+
+  /** Create a .tar.xz package file. The task will run universalStage() first, and then tarball and compress the stage
+    * output directory as the result.
+    */
+  def universalStagePackageTarXz: T[PathRef] = T {
+    val staged: PathRef = universalStage()
+    val out: Path = T.dest / (packageName() + universalTarXZExt())
+    Archiver().mkTarball(staged.path, out, universalTopLevelPath())(new XZCompressorOutputStream(_))
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
+
+  /** Create a .zip package file.
+    */
+  def universalPackageZip: T[PathRef] = T {
+    val zip = T.dest / (packageName() + ".zip")
+    ZipHelper.zip(universalPackageMappings(), zip)
+    T.log.info(s"Generated package: $zip")
+    PathRef(zip)
+  }
+
+  /** Create a .tar.zstd package file.
+    */
+  def universalPackageTarZstd: T[PathRef] = T {
+    val out: Path = T.dest / (packageName() + universalTarZstdExt())
+    Archiver().mkTarball(universalMappings(), out, universalTopLevelPath())(
+      new ZstdCompressorOutputStream(_, universalZstdCompressLevel())
+    )
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
+
+  /** Create a .tar.gz package file.
+    */
+  def universalPackageTarGzip: T[PathRef] = T {
+    val out: Path = T.dest / (packageName() + universalTarGZExt())
+    val parameter = new GzipParameters()
+    parameter.setCompressionLevel(universalGzipCompressLevel())
+    Archiver().mkTarball(universalMappings(), out, universalTopLevelPath())(new GzipCompressorOutputStream(_, parameter))
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
+
+  /** Create a .tar.bz2 package file.
+    */
+  def universalPackageTarBzip2: T[PathRef] = T {
+    val out: Path = T.dest / (packageName() + universalTarBzip2Ext())
+    Archiver().mkTarball(universalMappings(), out, universalTopLevelPath())(new BZip2CompressorOutputStream(_))
+    PathRef(out)
+  }
+
+  /** Create a .tar.xz package file.
+    */
+  def universalPackageTarXz: T[PathRef] = T {
+    val out: Path = T.dest / (packageName() + universalTarXZExt())
+    Archiver().mkTarball(universalMappings(), out, universalTopLevelPath())(new XZCompressorOutputStream(_))
+    T.log.info(s"Generated package: $out")
+    PathRef(out)
+  }
```

Comment:
or use the Apache Commons IOUtils.copy() instead.
or shall we just not care the JAVA 8 now?